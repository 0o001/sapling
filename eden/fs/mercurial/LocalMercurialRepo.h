/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include <folly/String.h>

namespace facebook {
namespace eden {

// Represents what we know about a mercurial repo on the local disk.
// We may add more information in here in the future.
// The thought is that we'll keep this repo checked out on the null
// rev, so that it is roughly equivalent to a bare checkout.
// See also LocalMercurialRepoAndRev
class LocalMercurialRepo {
  // The path to the directory that contains the .hg dir for the repo
  folly::fbstring localPath_;

 public:
  explicit LocalMercurialRepo(folly::StringPiece path);

  const folly::fbstring& getPath() const;
};
}
}
