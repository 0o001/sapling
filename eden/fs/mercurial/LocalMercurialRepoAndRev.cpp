/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#include "LocalMercurialRepoAndRev.h"
#include "MercurialFullManifest.h"

namespace facebook {
namespace eden {

LocalMercurialRepoAndRev::LocalMercurialRepoAndRev(
    folly::StringPiece rev, std::shared_ptr<LocalMercurialRepo> repo)
    : rev_(rev.fbstr()),
      repo_(repo),
      manifest_(MercurialFullManifest::parseManifest(*this)) {}

const folly::fbstring& LocalMercurialRepoAndRev::getRev() const { return rev_; }

std::shared_ptr<LocalMercurialRepo> LocalMercurialRepoAndRev::getRepo() const {
  return repo_;
}

MercurialFullManifest& LocalMercurialRepoAndRev::getManifest() {
  return *manifest_.get();
}
}
}
