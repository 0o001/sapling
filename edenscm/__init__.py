# Copyright Facebook, Inc. 2019
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.


def _fixsyspath():
    """Fix sys.path so core edenscm modules (edenscmnative, and 3rd party
    libraries) are in sys.path
    """
    # Do not expose those modules to edenscm.__dict__
    import sys
    import os

    dirname = os.path.dirname

    # __file__ is "hg/edenscm/__init__.py"
    # libdir is "hg/"
    # Do not follow symlinks (ex. do not use "realpath"). It breaks buck build.
    libdir = dirname(dirname(os.path.abspath(__file__)))

    # Make "edenscmdeps.zip" available in sys.path. It includes 3rd party
    # pure-Python libraries like IPython, thrift runtime, etc.
    #
    # Note: On Windows, the released version of hg uses python27.zip for all
    # pure Python modules including edenscm and everything in edenscmdeps.zip,
    # so not being able to locate edenscmdeps.zip is not fatal.
    for candidate in [libdir, os.path.join(libdir, "build")]:
        depspath = os.path.join(candidate, "edenscmdeps.zip")
        if os.path.exists(depspath) and depspath not in sys.path:
            sys.path.insert(0, depspath)

    # Make sure "edenscmnative" can be imported. Error early.
    import edenscmnative

    edenscmnative.__name__


_fixsyspath()


# Keep the module clean
del globals()["_fixsyspath"]
