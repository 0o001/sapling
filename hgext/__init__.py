# Copyright 2005 Mercurial Contributors
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.
