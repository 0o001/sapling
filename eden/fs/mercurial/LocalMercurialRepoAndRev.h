/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include "LocalMercurialRepo.h"
#include "MercurialFullManifest.h"

namespace facebook {
namespace eden {

class MercurialFullManifest;

// References a repo at a specific revision.
// Since we don't intend to have a full checkout maintained at any revision
// and we may also end up serving multiple users and checkouts, we need
// a way to reference the source of the history as well as a current
// revision.
// In addition to referencing the repo and revision, this provides
// an accessor to the manifest.
// At present we only have access to the full manifest information in
// mercurial, and we materialize this during construction.  The intention
// is that we'll do this all lazily in the future when hg has support
// for querying it in that fashion.
class LocalMercurialRepoAndRev
    : public std::enable_shared_from_this<LocalMercurialRepoAndRev> {
  folly::fbstring rev_;
  std::shared_ptr<LocalMercurialRepo> repo_;
  std::unique_ptr<MercurialFullManifest> manifest_;

 public:
  LocalMercurialRepoAndRev(folly::StringPiece rev,
                           std::shared_ptr<LocalMercurialRepo> repo);
  const folly::fbstring& getRev() const;
  std::shared_ptr<LocalMercurialRepo> getRepo() const;

  MercurialFullManifest& getManifest();
};
}
}
