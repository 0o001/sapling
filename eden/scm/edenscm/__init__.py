# Copyright (c) Facebook, Inc. and its affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

from __future__ import absolute_import


def _fixsyspath():
    """Fix sys.path so core edenscm modules (edenscmnative, and 3rd party
    libraries) are in sys.path
    """
    # Do not expose those modules to edenscm.__dict__
    import sys
    import os

    dirname = os.path.dirname

    # __file__ is "hg/edenscm/__init__.py"
    # libdir is "hg/"
    # Do not follow symlinks (ex. do not use "realpath"). It breaks buck build.
    libdir = dirname(dirname(os.path.abspath(__file__)))

    # Make "edenscmdeps.zip" available in sys.path. It includes 3rd party
    # pure-Python libraries like IPython, thrift runtime, etc.
    #
    # Note: On Windows, the released version of hg uses python27.zip for all
    # pure Python modules including edenscm and everything in edenscmdeps.zip,
    # so not being able to locate edenscmdeps.zip is not fatal.
    for candidate in [libdir, os.path.join(libdir, "build")]:
        depspath = os.path.join(candidate, "edenscmdeps.zip")
        if os.path.exists(depspath) and depspath not in sys.path:
            sys.path.insert(0, depspath)

    # Make sure "edenscmnative" can be imported. Error early.
    import edenscmnative

    edenscmnative.__name__


_fixsyspath()


# Keep the module clean
del globals()["_fixsyspath"]


def run(args=None, fin=None, fout=None, ferr=None):
    import sys

    if args is None:
        args = sys.argv

    if args[1:4] == ["serve", "--cmdserver", "chgunix2"]:
        # chgserver code path

        # no demandimport, since chgserver wants to preimport everything.
        from .mercurial import dispatch

        dispatch.runchgserver()
    else:
        # non-chgserver code path
        # - no chg in use: hgcommands::run -> HgPython::run_hg -> here
        # - chg client: chgserver.runcommand -> bindings.commands.run ->
        #               hgcommands::run -> HgPython::run_hg -> here

        from . import traceimport

        traceimport.enable()

        # enable demandimport after enabling traceimport
        from . import hgdemandimport

        hgdemandimport.enable()

        # demandimport has side effect on importing dispatch.
        # so 'import dispatch' happens after demandimport
        from .mercurial import dispatch

        dispatch.run(args, fin, fout, ferr)
