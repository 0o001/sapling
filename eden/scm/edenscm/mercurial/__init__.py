# Portions Copyright (c) Facebook, Inc. and its affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

# __init__.py - Startup and module loading logic for Mercurial.
#
# Copyright 2015 Gregory Szorc <gregory.szorc@gmail.com>
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

from __future__ import absolute_import

import os
import sys

# Allow 'from mercurial import demandimport' to keep working.
from edenscm import hgdemandimport


# pyre-fixme[11]: Annotation `hgdemandimport` is not defined as a type.
demandimport = hgdemandimport

__all__ = []

if getattr(sys, "platform") == "win32":
    configdir = os.path.join(
        getattr(os, "environ").get("PROGRAMDATA") or r"\ProgramData",
        "Facebook",
        "Mercurial",
    )
else:
    configdir = "/etc/mercurial"


def shoulduselegacy(name):
    legacy = getattr(os, "environ").get("HGLEGACY")
    if legacy is not None:
        return name in legacy.split()
    else:
        return os.path.lexists(os.path.join(configdir, "legacy.%s" % name))
