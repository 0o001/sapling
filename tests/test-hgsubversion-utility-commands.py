# no-check-code -- see T24862348

from __future__ import absolute_import

import os
import re

import test_hgsubversion_util
from edenscm.hgext import rebase
from edenscm.hgext.hgsubversion import (
    compathacks,
    svncommands,
    svnmeta,
    util,
    verify,
    wrappers,
)
from edenscm.mercurial import commands, context, hg, node, revlog, util as hgutil


expected_info_output = """URL: %(repourl)s/%(branch)s
Repository Root: %(repourl)s
Repository UUID: df2126f7-00ab-4d49-b42c-7e981dde0bcf
Revision: %(rev)s
Node Kind: directory
Last Changed Author: %(author)s
Last Changed Rev: %(rev)s
Last Changed Date: %(date)s
"""


def repourl(repo_path):
    return util.normalize_url(test_hgsubversion_util.fileurl(repo_path))


class UtilityTests(test_hgsubversion_util.TestBase):
    stupid_mode_tests = True

    def test_info_single(self, custom=False):
        if custom:
            subdir = None
            config = {"hgsubversionbranch.default": "trunk/"}
        else:
            subdir = "trunk"
            config = {}
        repo, repo_path = self.load_and_fetch(
            "two_heads.svndump", subdir=subdir, config=config
        )
        hg.update(self.repo, "tip")
        u = self.ui()
        u.pushbuffer()
        svncommands.info(u, self.repo)
        actual = u.popbuffer()
        expected = expected_info_output % {
            "date": "2008-10-08 01:39:29 +0000 (Wed, 08 Oct 2008)",
            "repourl": repourl(repo_path),
            "branch": "trunk",
            "rev": 6,
            "author": "durin",
        }
        self.assertMultiLineEqual(expected, actual)
        # let's set the config and test again
        self.commitchanges(
            changes=[("a", "a", "random string")],
            extras={
                "global_rev": 123456789,
                "convert_revision": "svn:df2126f7-00ab-4d49-b42c-7e981dde0bcf/trunk@7",
            },
        )
        hg.update(self.repo, "tip")
        svncommands.updatemeta(u, self.repo, [])
        u.pushbuffer()
        u.setconfig("hgsubversion", "useglobalrevinsvninfo", True)
        svncommands.info(u, self.repo)
        actual = u.popbuffer()
        expected = expected_info_output % {
            "date": "2008-10-07 20:59:48 -0500 (Tue, 07 Oct 2008)",
            "repourl": repourl(repo_path),
            "branch": "trunk",
            "rev": 123456789,
            "author": "an_author",
        }
        self.assertMultiLineEqual(expected, actual)

    def test_info_custom_single(self):
        self.test_info_single(custom=True)

    def test_missing_metadata(self):
        self._load_fixture_and_fetch("two_heads.svndump")
        os.remove(self.repo.sharedvfs.join("svn/branch_info"))
        svncommands.updatemeta(self.ui(), self.repo, [])

        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        self.assertRaises(hgutil.Abort, self.repo.svnmeta)
        self.assertRaises(
            hgutil.Abort, svncommands.info, self.ui(), repo=self.repo, args=[]
        )
        self.assertRaises(
            hgutil.Abort, svncommands.genignore, self.ui(), repo=self.repo, args=[]
        )

        os.remove(self.repo.localvfs.join("hgrc"))
        self.assertRaises(hgutil.Abort, self.repo.svnmeta)
        self.assertRaises(
            hgutil.Abort, svncommands.info, self.ui(), repo=self.repo, args=[]
        )
        self.assertRaises(
            hgutil.Abort, svncommands.genignore, self.ui(), repo=self.repo, args=[]
        )

        self.assertRaises(
            hgutil.Abort, svncommands.rebuildmeta, self.ui(), repo=self.repo, args=[]
        )

    def test_rebase(self):
        self._load_fixture_and_fetch("two_revs.svndump")
        parents = (self.repo[0].node(), revlog.nullid)

        def filectxfn(repo, memctx, path):
            return compathacks.makememfilectx(
                repo,
                memctx=memctx,
                path=path,
                data="added",
                islink=False,
                isexec=False,
                copied=False,
            )

        lr = self.repo
        ctx = context.memctx(
            lr,
            parents,
            "automated test",
            ["added_bogus_file", "other_added_file"],
            filectxfn,
            "testy",
            "2008-12-21 16:32:00 -0500",
            {"branch": "localbranch"},
        )
        lr.commitctx(ctx)
        self.assertEqual(self.repo["tip"].branch(), "localbranch")
        beforerebasehash = self.repo["tip"].node()
        hg.update(self.repo, "tip")
        wrappers.rebase(rebase.rebase, self.ui(), self.repo, svn=True)
        self.assertEqual(self.repo["tip"].branch(), "localbranch")
        self.assertEqual(self.repo["tip"].parents()[0].parents()[0], self.repo[0])
        self.assertNotEqual(beforerebasehash, self.repo["tip"].node())

    def test_genignore(self, layout="auto"):
        """ Test generation of .hgignore file. """
        if layout == "custom":
            config = {"hgsubversionbranch.default": "trunk"}
        else:
            config = {}
        repo = self._load_fixture_and_fetch(
            "ignores.svndump", layout=layout, noupdate=False, config=config
        )
        u = self.ui()
        u.pushbuffer()
        svncommands.genignore(u, repo, self.wc_path)
        self.assertMultiLineEqual(
            open(os.path.join(self.wc_path, ".hgignore")).read(),
            ".hgignore\nsyntax:glob\nblah\notherblah\nbaz/magic\n",
        )

    def test_genignore_single(self):
        self.test_genignore(layout="single")

    def test_genignore_custom(self):
        self.test_genignore(layout="custom")

    def test_list_authors(self):
        repo_path = self.load_svndump("replace_trunk_with_branch.svndump")
        u = self.ui()
        u.pushbuffer()
        svncommands.listauthors(
            u, args=[test_hgsubversion_util.fileurl(repo_path)], authors=None
        )
        actual = u.popbuffer()
        self.assertMultiLineEqual(actual, "Augie\nevil\n")

    def test_list_authors_map(self):
        repo_path = self.load_svndump("replace_trunk_with_branch.svndump")
        author_path = os.path.join(repo_path, "authors")
        svncommands.listauthors(
            self.ui(),
            args=[test_hgsubversion_util.fileurl(repo_path)],
            authors=author_path,
        )
        self.assertMultiLineEqual(open(author_path).read(), "Augie=\nevil=\n")

    def test_svnverify(self):
        repo, repo_path = self.load_and_fetch("binaryfiles.svndump", noupdate=False)
        ret = verify.verify(self.ui(), repo, [], rev=1)
        self.assertEqual(0, ret)
        repo_path = self.load_svndump("binaryfiles-broken.svndump")
        u = self.ui()
        u.pushbuffer()
        ret = verify.verify(u, repo, [test_hgsubversion_util.fileurl(repo_path)], rev=1)
        output = u.popbuffer()
        self.assertEqual(1, ret)
        output = re.sub(r"file://\S+", "file://", output)
        self.assertMultiLineEqual(
            """\
verifying d51f46a715a1 against file://
difference in: binary2
unexpected file: binary1
missing file: binary3
""",
            output,
        )

    def test_corruption(self):
        SUCCESS = 0
        FAILURE = 1

        repo, repo_path = self.load_and_fetch(
            "correct.svndump", layout="single", subdir=""
        )

        ui = self.ui()

        self.assertEqual(SUCCESS, verify.verify(ui, self.repo, rev="tip"))

        corrupt_source = test_hgsubversion_util.fileurl(
            self.load_svndump("corrupt.svndump")
        )

        repo.ui.setconfig("paths", "default", corrupt_source)

        ui.pushbuffer()
        code = verify.verify(ui, repo, rev="tip")
        actual = ui.popbuffer()

        actual = actual.replace(corrupt_source, "$REPO")
        actual = set(actual.splitlines())

        expected = set(
            [
                "verifying 78e965230a13 against $REPO@1",
                "missing file: missing-file",
                "wrong flags for: executable-file",
                "wrong flags for: symlink",
                "wrong flags for: regular-file",
                "difference in: another-regular-file",
                "difference in: regular-file",
                "unexpected file: empty-file",
            ]
        )

        self.assertEqual((FAILURE, expected), (code, actual))

    def test_svnrebuildmeta(self):
        otherpath = self.load_svndump("binaryfiles-broken.svndump")
        otherurl = test_hgsubversion_util.fileurl(otherpath)
        _, repopath = self.load_and_fetch("replace_trunk_with_branch.svndump")
        repourl = test_hgsubversion_util.fileurl(repopath)
        # rebuildmeta with original repo
        svncommands.rebuildmeta(self.ui(), repo=self.repo, args=[])
        # rebuildmeta with original repo explicitly
        svncommands.rebuildmeta(self.ui(), repo=self.repo, args=[repourl])
        # rebuildmeta with unrelated repo
        self.assertRaises(
            hgutil.Abort,
            svncommands.rebuildmeta,
            self.ui(),
            repo=self.repo,
            args=[otherurl],
        )
        # rebuildmeta --unsafe-skip-uuid-check with unrelated repo
        svncommands.rebuildmeta(
            self.ui(), repo=self.repo, args=[otherurl], unsafe_skip_uuid_check=True
        )

    def test_svntruncatemeta(self):
        repo, repo_path = self.load_and_fetch("two_heads.svndump", config={})
        orig_tip_rev = len(repo) - 1
        orig_tip_svn_rev = util.getsvnrevnum(repo["tip"])

        orig_tip = repo["tip"]

        # Create a new commit that has the same svnrev as the old one
        new_ctx = context.metadataonlyctx(
            repo, orig_tip, text="foo", extra=orig_tip.extra()
        )
        new_tip = repo[repo.commitctx(new_ctx)]

        # Before the truncate, the svnrev points at the old commit
        self.assertEqual(
            list(repo.revs("svnrev(%s)" % orig_tip_svn_rev)), [orig_tip_rev]
        )

        svncommands.truncatemeta(self.ui(), repo=self.repo, args=[])
        repo.invalidate()

        # After the truncate, the svnrev points at the new commit
        new_tip_rev = new_tip.rev()
        self.assertEqual(
            list(repo.revs("svnrev(%s)" % orig_tip_svn_rev)), [new_tip_rev]
        )

    def test_svn_reposubdir_config(self):
        otherpath = self.load_svndump("binaryfiles-broken.svndump")
        otherurl = test_hgsubversion_util.fileurl(otherpath)
        _, repopath = self.load_and_fetch(
            "subdir_is_file_prefix.svndump", subdir="flaf"
        )
        repourl = test_hgsubversion_util.fileurl(repopath)
        ui = self.ui()

        # rebuildmeta with original repo
        svncommands.rebuildmeta(ui, repo=self.repo, args=[repourl])
        old_subdir = svnmeta.SVNMeta(self.repo).subdir
        # remove metadata, rebuild with a wrong subdir config
        # this should raise
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        ui.setconfig("hgsubversion", "reposubdir", "new_wrong_subdir")
        self.assertRaises(
            AssertionError,
            svncommands.rebuildmeta,
            ui=ui,
            repo=self.repo,
            args=[repourl],
        )
        # remove metadata, rebuild with the correct subdir config
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        ui.setconfig("hgsubversion", "reposubdir", old_subdir)
        svncommands.rebuildmeta(ui, repo=self.repo, args=[repourl])
        new_subdir = svnmeta.SVNMeta(self.repo).subdir
        self.assertEqual(old_subdir, new_subdir)
        # remove metadata, rebuild with the correct subdir config, with a unrealted url
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        ui.setconfig("hgsubversion", "reposubdir", old_subdir)
        ui.setconfig("hgsubversion", "repouuid", "924a052a-5e5a-4a8e-a677-da5565bec340")
        svncommands.rebuildmeta(ui, repo=self.repo, args=[otherurl])
        another_subdir = svnmeta.SVNMeta(self.repo).subdir
        self.assertEqual(old_subdir, another_subdir)

    def test_svn_repouuid_config(self):
        otherpath = self.load_svndump("binaryfiles-broken.svndump")
        otherurl = test_hgsubversion_util.fileurl(otherpath)
        _, repopath = self.load_and_fetch("replace_trunk_with_branch.svndump")
        repourl = test_hgsubversion_util.fileurl(repopath)
        # rebuildmeta with original repo
        svncommands.rebuildmeta(self.ui(), repo=self.repo, args=[])
        # updatemeta with the wrong, pre-configured uuid
        ui = self.ui()
        ui.setconfig("hgsubversion", "repouuid", "a wrong uuid")
        self.assertRaises(
            hgutil.Abort, svncommands.updatemeta, ui=ui, repo=self.repo, args=[repourl]
        )
        # updatemeta with the right, pre-configured uuid
        ui.setconfig("hgsubversion", "repouuid", "5b65bade-98f3-4993-a01f-b7a6710da339")
        svncommands.rebuildmeta(ui, repo=self.repo, args=[repourl])
        # updatemeta with the right, pre-configured uuid, and unrelated repo
        ui.setconfig("hgsubversion", "repouuid", "5b65bade-98f3-4993-a01f-b7a6710da339")
        svncommands.rebuildmeta(ui, repo=self.repo, args=[otherurl])
        # updatemeta with empty meta
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        ui.setconfig("hgsubversion", "repouuid", "5b65bade-98f3-4993-a01f-b7a6710da339")
        svncommands.rebuildmeta(ui, repo=self.repo, args=[otherurl])

    def test_svn_repouuid_reposuburl_config_not_accessing_svn(self):
        _, repopath = self.load_and_fetch(
            "subdir_is_file_prefix.svndump", subdir="flaf"
        )
        repourl = test_hgsubversion_util.fileurl(repopath)
        ui = self.ui()

        # rebuildmeta with original repo
        svncommands.rebuildmeta(ui, repo=self.repo, args=[repourl])
        subdir = svnmeta.SVNMeta(self.repo).subdir
        uuid = svnmeta.SVNMeta(self.repo).uuid
        # remove metadata, rebuild without configs.
        # In attempt to use the some_random_string as url, this should raise.
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        self.assertRaises(
            hgutil.Abort,
            svncommands.rebuildmeta,
            ui=ui,
            repo=self.repo,
            args=["svn+ssh://some_random_string"],
        )
        # remove metadata, rebuild with the correct configs
        # This time it doesn't try to use the wrong url, this should pass.
        test_hgsubversion_util.rmtree(self.repo.sharedvfs.join("svn"))
        ui.setconfig("hgsubversion", "reposubdir", subdir)
        ui.setconfig("hgsubversion", "repouuid", uuid)
        svncommands.rebuildmeta(
            ui=ui, repo=self.repo, args=["svn+ssh://some_random_string"]
        )


if __name__ == "__main__":
    import silenttestrunner

    silenttestrunner.main(__name__)
