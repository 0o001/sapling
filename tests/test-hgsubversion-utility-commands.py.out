verifying 78e965230a13 against file:///[\w/]*/testrepo-1@1 (re)
verifying 78e965230a13 against file:///[\w/]*/testrepo-1@1 (re)
verifying d51f46a715a1 against file:///[\w/]*/testrepo-1/trunk@3 (re)
verifying d51f46a715a1 against file:///[\w/]*/testrepo-1/trunk@3 (re)
